From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: "Sofiane H. Djerbi" <46628754+kugge@users.noreply.github.com>
Date: Fri, 10 Feb 2023 17:45:59 +0200
Subject: [PATCH] Kaiiju Configuration


diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index 8d51301e1a0a2a528353b90e2e2d0e4c69c6a4f5..d8d06160268a977fcd933e0e4dbc8d5c39af78f0 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -2198,6 +2198,13 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
         }
         // Paper end
 
+        // Kaiiju start
+        @NotNull
+        public org.bukkit.configuration.file.YamlConfiguration getKaiijuConfig() {
+            throw new UnsupportedOperationException("Not supported yet.");
+        }
+        // Kaiiju end
+
         /**
          * Sends the component to the player
          *
