From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Xymb <xymb@endcrystal.me>
Date: Mon, 22 Jan 2024 01:20:04 +0100
Subject: [PATCH] WIP: Disable fluid ticking.


diff --git a/src/main/java/dev/kaiijumc/kaiiju/KaiijuConfig.java b/src/main/java/dev/kaiijumc/kaiiju/KaiijuConfig.java
index 3ce098436d01146316bd76c76fa9d5e6b3d6f819..3a8240a8fe9c90ff3fd9bfdd8e341b178037960d 100644
--- a/src/main/java/dev/kaiijumc/kaiiju/KaiijuConfig.java
+++ b/src/main/java/dev/kaiijumc/kaiiju/KaiijuConfig.java
@@ -237,7 +237,11 @@ public class KaiijuConfig {
     // About 8.3% of tick time with 100 bots on a pregenerated world
     public static boolean throttleLeavesTicking = false;
 
+    // WIP for development
+    public static boolean disableFluidTicking = false;
+
     private static void optimizationsDirtyWillChangeName() {
         throttleLeavesTicking = getBoolean("optimizations-dirty-will-change-name.throttle-leaves-ticking", throttleLeavesTicking);
+        disableFluidTicking = getBoolean("optimizations-dirty-will-change-name.disable-fluid-ticking", disableFluidTicking);
     }
 }
diff --git a/src/main/java/net/minecraft/world/level/material/FluidState.java b/src/main/java/net/minecraft/world/level/material/FluidState.java
index 27f136815afc360387704fa1f2773e3816cccff6..a6fdbdccc1830ef2005dbfdda40df462fad8d073 100644
--- a/src/main/java/net/minecraft/world/level/material/FluidState.java
+++ b/src/main/java/net/minecraft/world/level/material/FluidState.java
@@ -77,6 +77,7 @@ public final class FluidState extends StateHolder<Fluid, FluidState> {
     }
 
     public void tick(Level world, BlockPos pos) {
+        if (dev.kaiijumc.kaiiju.KaiijuConfig.disableFluidTicking) return; // Kaiiju
         this.getType().tick(world, pos, this);
     }
 
